File: leetcode/2.cpp
#include <bits/stdc++.h>

using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode fake = ListNode(-1);
        ListNode *cur = &fake;
        int carry = 0;

        while (l1 != nullptr && l2 != nullptr)
        {
            int val = l1->val + l2->val + carry;
            cur->next = new ListNode(val % 10);
            carry = val / 10;
            cur = cur->next;
            l1 = l1->next;
            l2 = l2->next;
        }

        while (l1 != nullptr)
        {
            int val = l1->val + carry;
            cur->next = new ListNode(val % 10);
            carry = val / 10;
            cur = cur->next;
            l1 = l1->next;
        }

        while (l2 != nullptr)
        {
            int val = l2->val + carry;
            cur->next = new ListNode(val % 10);
            carry = val / 10;
            cur = cur->next;
            l2 = l2->next;
        }

        if (carry > 0)
        {
            cur->next = new ListNode(carry);
        }

        return fake.next;
    }
};

class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode fake = ListNode(-1);
        ListNode *cur = &fake;
        int carry = 0;

        while (l1 != nullptr || l2 != nullptr || carry)
        {
            int val = l1 != nullptr ? l1->val : 0;
            val += l2 != nullptr ? l2->val : 0;
            val += carry;
            cur->next = new ListNode(val % 10);
            carry = val / 10;
            cur = cur->next;
            l1 = l1 != nullptr ? l1->next : nullptr;
            l2 = l2 != nullptr ? l2->next : nullptr;
        }
        return fake.next;
    }
};

File: python/leetcode/2.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
from typing import Optional


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        res = ListNode()
        node = res
        carry = 0
        while l1 or l2:
            v1, v2 = l1.val if l1 else 0, l2.val if l2 else 0
            v = v1 + v2 + carry
            c, v = divmod(v,10)
            node.next = ListNode(v)
            node = node.next
            carry = c
            l1, l2 = l1.next if l1 else None, l2.next if l2 else None
        if carry:
            node.next = ListNode(carry)
        return res.next